cupy/util.pyx

```diff
@@ -26,6 +29,16 @@ except AttributeError:  # python <3.3
 cdef list _memos = []
 
 
+cpdef intptr_t malloc(size_t size):
+    """Allocates a memory of specified size."""
+    return <intptr_t>mem.PyMem_Malloc(size)
+
+
+cpdef void free(intptr_t ptr):
+    """Frees the specified memory."""
+    mem.PyMem_Free(<void*>ptr)
```

Comment:
I am not sure I like this...It's a think wrapper, yes, but still incurs some overhead. Is there a good reason we can't simply call `PyMem_Malloc` and `PyMem_Free` wherever needed?

Reply:
Thank you for a comment.
As `PyMem_*` are `cdef` functions, they are not directly usable from `.py` files.
As for the overhead, when called from cython code, I suppose the call overhead is negligible compared to allocation itself. Not tested though.

Reply:
Another concern of this API is that it'd expose raw malloc and free to end users, which could go wild...

Reply:
We commented at the same time 😛 

> As `PyMem_*` are `cdef` functions, they are not directly usable from `.py` files.

True. So for cuTENSOR it's not possible/too difficult to manage memory within `.pyx`?

Reply:
It's possible and not difficult either.
Actually that's another option to consider.
It's like moving `class Handle` to `cutensor.pyx`

```py
# cutensor.pyx

cdef class Handle:
    cef intptr_t ptr
    def __init__(self):
        self.ptr = PyMem_Malloc(...)
    def __del__(self):
        PyMem_Free(self.ptr)
    
cpdef Handle init():
    handle = Handle()
    cutensorInit(handle.ptr)
    return handle
```

Reply:
I'll make another PR with this version.

Reply:
Well, I haven't implemented it yet, but I found one disadvantage of that version.
That is, it will be unable to recycle the memory allocated for handles/descriptors.

I don't think this is such a great disadvantage, but this kind of common optimization will be unable to do.

Reply:
This is yet another candidate, which does not have any of the aforementioned problems.

```py
# cupy/cutensor.py

handle = cutensor.Handle()
cutensor.init(handle)
```

```py
# cupy/cuda/cutensor.pyx

cdef class Handle:  # same as the previous version
    cdef intptr_t ptr
    def __init__(self):
        self.ptr = PyMem_Malloc(...)
    def __del__(self):
        PyMem_Free(self.ptr)
    
cpdef void init(Handle):
    cutensorInit(handle.ptr)
```

Reply:
To be honest I am not sure what disadvantage was referred to. The two `*.pyx` implementations look good to me.

Reply:
In version 3 above, `malloc` is inevitable whenever initializing handles/descriptors/etc., which is not the case in C API and version 4.